File: perfbench-code/wicket15-jpa/src/main/java/wicketjpa/wicket/EditBorder.java
package wicketjpa.wicket;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import javax.persistence.Entity;

import org.apache.wicket.Component;
import org.apache.wicket.ajax.AjaxRequestTarget;
import org.apache.wicket.ajax.form.AjaxFormComponentUpdatingBehavior;
import org.apache.wicket.behavior.Behavior;
import org.apache.wicket.feedback.ContainerFeedbackMessageFilter;
import org.apache.wicket.markup.ComponentTag;
import org.apache.wicket.markup.html.border.Border;
import org.apache.wicket.markup.html.form.FormComponent;
import org.apache.wicket.markup.html.panel.FeedbackPanel;
import org.apache.wicket.model.CompoundPropertyModel;
import org.apache.wicket.model.IModel;
import org.apache.wicket.validation.IValidatable;
import org.apache.wicket.validation.IValidator;
import org.apache.wicket.validation.ValidationError;
import org.hibernate.validator.ClassValidator;
import org.hibernate.validator.InvalidValue;

public class EditBorder extends Border {
    
    private static final Map<Class, ClassValidator> classValidatorCache =
            new ConcurrentHashMap<Class, ClassValidator>();

    public EditBorder(String id) {
        super(id);
        addToBorder(new FeedbackPanel("message", new ContainerFeedbackMessageFilter(this)));
    }

    public EditBorder(String id, FormComponent fc, boolean ajax) {
        this(id);
        add(fc, ajax);
    }

    public EditBorder(String id, FormComponent fc) {
        this(id, fc, false);
    }

    public void add(FormComponent fc) {
        add(fc, false);
    }

    public void add(final FormComponent fc, boolean ajax) {
        super.add(fc);
        fc.add(new Behavior() {
	        @Override
	        public void onConfigure(Component component)
	        {
		        super.onConfigure(component);
                IModel model = component.getInnermostModel();
                if (model != null && model instanceof CompoundPropertyModel) {
                    CompoundPropertyModel cpm = (CompoundPropertyModel) model;
                    Class clazz = cpm.getObject().getClass();
                    if (clazz.isAnnotationPresent(Entity.class)) {
                        fc.add(getValidator(clazz, fc.getId()));
                    }
                }
            }
        });
        add(new Behavior() {
            @Override
            public void onComponentTag(Component c, ComponentTag tag) {
                if (!fc.isValid()) {
                    tag.put("class", "input errors");
                }
            } 
        });
        if(ajax) {
            setOutputMarkupId(true);
            fc.add(new AjaxFormComponentUpdatingBehavior("onblur") {
                @Override
                protected void onUpdate(AjaxRequestTarget target) {
                    getFormComponent().validate();
                    target.add(EditBorder.this);
                }
                @Override
                protected void onError(AjaxRequestTarget target, RuntimeException e) {
                    target.add(EditBorder.this);
                }
            });
        }
    }

    private IValidator getValidator(final Class clazz, final String expression) {
        return new IValidator() {
            public void validate(IValidatable v) {
                ClassValidator cv = classValidatorCache.get(clazz);
                if (cv == null) {
                    cv = new ClassValidator(clazz);
                    classValidatorCache.put(clazz, cv);
                }
                InvalidValue[] invalidValues = cv.getPotentialInvalidValues(expression, v.getValue());
                if (invalidValues.length > 0) {
                    String message = invalidValues[0].getMessage();                    
                    v.error(new ValidationError().setMessage(message));
                }
            }
        };
    }
}

File: perfbench-code/swicket-jpa/src/main/scala/swicketjpa/wicket/EditBorder.scala
package swicketjpa.wicket

import javax.persistence.Entity
import java.util.concurrent.ConcurrentHashMap
import org.apache.wicket.Component
import org.apache.wicket.ajax.AjaxRequestTarget
import org.apache.wicket.ajax.form.AjaxFormComponentUpdatingBehavior
import org.apache.wicket.behavior.AbstractBehavior
import org.apache.wicket.feedback.ContainerFeedbackMessageFilter
import org.apache.wicket.markup.ComponentTag
import org.apache.wicket.markup.html.border.Border
import org.apache.wicket.markup.html.form.FormComponent
import org.apache.wicket.markup.html.panel.FeedbackPanel
import org.apache.wicket.model.CompoundPropertyModel
import org.apache.wicket.validation._
import org.hibernate.validator.ClassValidator

object EditBorder {
  val classValidatorCache = new ConcurrentHashMap[Class[_], ClassValidator[_]]()
}

class EditBorder(id: String) extends Border(id) {

  add(new FeedbackPanel("message", new ContainerFeedbackMessageFilter(this)))

  def this(id: String, fc: FormComponent[_], ajax: Boolean) = {
    this(id)
    add(fc, ajax)
  }

  def this(id: String, fc: FormComponent[_]) = this(id, fc, false)

  def add(fc: FormComponent[_]): Unit = add(fc, false)

  def add[T](fc: FormComponent[T], ajax: Boolean) = {
    super.add(fc)
    fc.add(new AbstractBehavior() {
      override def beforeRender(c: Component) = {
        super.beforeRender(c)
        val model = c.getInnermostModel()
        if (model != null && model.isInstanceOf[CompoundPropertyModel[_]]) {
          val cpm = model.asInstanceOf[CompoundPropertyModel[AnyRef]]
          val clazz = cpm.getObject.getClass          
          if (clazz.isAnnotationPresent(classOf[Entity])) {            
            fc.add(getValidator[T](clazz, fc.getId()))
          }
        }
      }
    })
    super.add(new AbstractBehavior() {
      override def onComponentTag(c: Component, tag: ComponentTag) =
        if (!fc.isValid()) tag.put("class", "input errors")
    })
    if(ajax) {
      setOutputMarkupId(true)
      fc.add(new AjaxFormComponentUpdatingBehavior("onblur") {
        override def onUpdate(target: AjaxRequestTarget) = {
          getFormComponent().validate()
          target.addComponent(EditBorder.this)
        }
        override def onError(target: AjaxRequestTarget, e: RuntimeException) =
          target.addComponent(EditBorder.this)        
      })
    }
  }

  def getValidator[T](clazz: Class[_], expression: String) = {
    new IValidator[T]() {
      override def validate(v: IValidatable[T]) = {
        var cv = EditBorder.classValidatorCache.get(clazz)
        if (cv == null) {
          cv = new ClassValidator(clazz)
          EditBorder.classValidatorCache.put(clazz, cv)
        }
        val invalidValues = cv.getPotentialInvalidValues(expression, v.getValue())        
        if (invalidValues.length > 0) {
          val message = invalidValues(0).getMessage()
          v.error(new ValidationError().setMessage(message))
        }
      }
    }
  }

}
